beacon-chain/forkchoice/doubly-linked-tree/store.go

```diff
@@ -0,0 +1,234 @@
+		if !ok || currentNode == nil {
+			return errInvalidProposerBoostRoot
+		}
+		proposerScore, err = computeProposerBoostScore(newBalances)
+		if err != nil {
+			return err
+		}
+		currentNode.balance += proposerScore
+	}
+	s.previousProposerBoostRoot = s.proposerBoostRoot
+	s.previousProposerBoostScore = proposerScore
+	return nil
+}
+
+// NodeNumber returns the current number of nodes in the Store
+func (s *Store) NodeNumber() int {
+	return len(s.nodeByRoot)
+}
+
+// ProposerBoost of fork choice store.
+func (s *Store) proposerBoost() [fieldparams.RootLength]byte {
+	s.proposerBoostLock.RLock()
+	defer s.proposerBoostLock.RUnlock()
+	return s.proposerBoostRoot
+}
+
+// PruneThreshold of fork choice store.
+func (s *Store) PruneThreshold() uint64 {
+	return s.pruneThreshold
+}
+
+// head starts from justified root and then follows the best descendant links
+// to find the best block for head. This function assumes a lock on s.nodesLock
+func (s *Store) head(ctx context.Context, justifiedRoot [32]byte) ([32]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "doublyLinkedForkchoice.head")
+	defer span.End()
+
+	// JustifiedRoot has to be known
+	justifiedNode, ok := s.nodeByRoot[justifiedRoot]
+	if !ok || justifiedNode == nil {
+		return [32]byte{}, errUnknownJustifiedRoot
+	}
+
+	// If the justified node doesn't have a best descendant,
+	// the best node is itself.
+	bestDescendant := justifiedNode.bestDescendant
+	if bestDescendant == nil {
+		bestDescendant = justifiedNode
+	}
+
+	if !bestDescendant.viableForHead(s.justifiedEpoch, s.finalizedEpoch) {
+		return [32]byte{}, fmt.Errorf("head at slot %d with weight %d is not eligible, finalizedEpoch %d != %d, justifiedEpoch %d != %d",
+			bestDescendant.slot, bestDescendant.weight/10e9, bestDescendant.finalizedEpoch, s.finalizedEpoch, bestDescendant.justifiedEpoch, s.justifiedEpoch)
+	}
+
+	// Update metrics.
+	if bestDescendant != s.headNode {
+		headChangesCount.Inc()
+		headSlotNumber.Set(float64(bestDescendant.slot))
+		s.headNode = bestDescendant
+	}
+
+	return bestDescendant.root, nil
+}
+
+// insert registers a new block node to the fork choice store's node list.
+// It then updates the new node's parent with best child and descendant node.
+func (s *Store) insert(ctx context.Context,
+	slot types.Slot,
+	root, parentRoot [fieldparams.RootLength]byte,
+	justifiedEpoch, finalizedEpoch types.Epoch, optimistic bool) error {
+	_, span := trace.StartSpan(ctx, "doublyLinkedForkchoice.insert")
+	defer span.End()
+
+	s.nodesLock.Lock()
+	defer s.nodesLock.Unlock()
+
+	// Return if the block has been inserted into Store before.
+	if _, ok := s.nodeByRoot[root]; ok {
+		return nil
+	}
+
+	parent := s.nodeByRoot[parentRoot]
+
+	n := &Node{
+		slot:           slot,
+		root:           root,
+		parent:         parent,
+		justifiedEpoch: justifiedEpoch,
+		finalizedEpoch: finalizedEpoch,
+		optimistic:     optimistic,
+	}
+
+	s.nodeByRoot[root] = n
+	if parent != nil {
```

Comment:
I think we can treat `parent != nil` and `s.treeRootNode == nil` as the same condition and simplify to
```go
if parent != nil {
		parent.children = append(parent.children, n)
		if err := s.treeRootNode.updateBestDescendant(ctx, s.justifiedEpoch, s.finalizedEpoch); err != nil {
			return err
		}
	} else {
		s.treeRootNode = n
		s.headNode = n
	}
```

Reply:
Here it's not that clear to me. The caller can pass here an unkonwn parent root here. I followed the current implementation in the protoarray and did not want to change it. Refactoring this is not simplified as in your example cause we have to return an error if the parent is nil and the treeRootNode is not (to avoid having two nodes with a nil parent).